File: Task/CUSIP/C-sharp/cusip.cs
using System;
using System.Collections.Generic;

namespace CUSIP {
    class Program {
        static bool IsCusip(string s) {
            if (s.Length != 9) return false;
            int sum = 0;
            for (int i = 0; i <= 7; i++) {
                char c = s[i];

                int v;
                if (c >= '0' && c <= '9') {
                    v = c - 48;
                }
                else if (c >= 'A' && c <= 'Z') {
                    v = c - 55;  // lower case letters apparently invalid
                }
                else if (c == '*') {
                    v = 36;
                }
                else if (c == '#') {
                    v = 38;
                }
                else {
                    return false;
                }
                if (i % 2 == 1) v *= 2;  // check if odd as using 0-based indexing
                sum += v / 10 + v % 10;
            }
            return s[8] - 48 == (10 - (sum % 10)) % 10;
        }

        static void Main(string[] args) {
            List<string> candidates = new List<string>() {
                "037833100",
                "17275R102",
                "38259P508",
                "594918104",
                "68389X106",
                "68389X105"
            };
            foreach (var candidate in candidates) {
                Console.WriteLine("{0} -> {1}", candidate, IsCusip(candidate) ? "correct" : "incorrect");
            }
        }
    }
}

File: Task/CUSIP/PascalABC.NET/cusip.pas
function IsCusip(s: string): boolean;
begin
  result := false;
  if (s.Length <> 9) then exit;
  var sum := 0;
  var v := 0;
  for var i := 1 to 8 do
  begin
    var c := s[i];
    if c.isdigit then
      v := c.todigit
    else if (c >= 'A') and (c <= 'Z') then
      v := ord(c) - ord('A') + 10
    else if c = '*' then
      v := 36
    else if c = '@' then
      v := 37
    else if c = '#' then
      v := 38;
    if (i mod 2) = 0 then
      v := v * 2;
    sum += (v div 10) + (v mod 10);
  end;
  var check := (10 - (sum mod 10)) mod 10;
  result := s[9].todigit = check;
end;

begin
  var candidates := |'037833100', '17275R102', '38259P508', '594918104', '68389X106', '68389X105'|;
  foreach var candidate in candidates do
    println(candidate, if IsCusip(candidate) then 'correct' else 'incorrect');
end.
